src/main/java/pl/edu/agh/firecell/engine/algorithm/BasicAlgorithm.java

```diff
@@ -4,80 +4,139 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import pl.edu.agh.firecell.model.Cell;
+import pl.edu.agh.firecell.model.Material;
+import pl.edu.agh.firecell.model.MatterState;
 import pl.edu.agh.firecell.model.State;
 import pl.edu.agh.firecell.model.util.NeighbourUtils;
 
+import java.util.Optional;
+
 
 public class BasicAlgorithm implements Algorithm {
 
     private final Logger logger = LoggerFactory.getLogger(this.getClass());
-    private final double deltaTime;
-    public static final double CONVECTION_COEFFICIENT = 0.1;
+    private final TemperatureCalculation temperatureCalculation;
+    public static final double CONVECTION_COEFFICIENT = 1;
     // should be dependent on the material in the future
-    public static final double CONDUCTIVITY_COEFFICIENT = 0.1;
-    public static final int MAX_BURNING_TIME = 5;
-    public static final double MIN_BURNING_TEMPERATURE = 100;
+    public static final double CONDUCTIVITY_COEFFICIENT = 0.2;
+    public static final int MAX_BURNING_TIME = 50;
+
 
     public BasicAlgorithm(double deltaTime) {
-        this.deltaTime = deltaTime;
+        this.temperatureCalculation = new TemperatureCalculation(deltaTime);
     }
 
     @Override
     public Cell compute(State oldState, Vector3i cellIndex) {
 
         Cell oldCell = oldState.getCell(cellIndex);
 
-        // temperature propagation
-        double newTemperature = computeNewTemperature(oldState, cellIndex, oldCell);
-
-        // fire propagation
+        double newTemperature = temperatureCalculation.computeNewTemperature(oldState, cellIndex, oldCell);
         boolean newFlammable = oldCell.flammable();
         int newBurningTime = oldCell.burningTime();
-        if (oldCell.flammable() && oldCell.burningTime() >= 0 && newTemperature > MIN_BURNING_TEMPERATURE) {
-            newBurningTime++;
-        }
-        if (newBurningTime > MAX_BURNING_TIME) {
-            newFlammable = false;
+        int newRemainingHeightOfFirePillar = oldCell.remainingHeightOfFirePillar();
+        boolean newPossibleToGoUp = oldCell.possibleToGoUp();
```

Comment:
Assigning old values to new values is counterintuitive, especially if you process and compute them later. Just use old values during computation, it will make it more readable and bug-safe